grammar.js

```diff
@@ -1152,5 +1196,428 @@ module.exports = grammar({
       token.immediate('>'),
       alias(token.immediate(/[\w|\d]+/), $.name),
     ),
+
+    /**************************************************************************/
+    /* PlusCal: written inside block comments and transpiled into TLA+.       */
+    /* According to the BNF from p. 60-62 of both p-manual and c-manual       */
+    /* (https://lamport.azurewebsites.net/tla/p-manual.pdf)                   */
+    /* (https://lamport.azurewebsites.net/tla/c-manual.pdf)                   */
+    /**************************************************************************/
+
+    _pcal_algorithm: $ => choice(
```

Comment:
Instead of having two parallel sets of rules for the different styles, can we combine the styles in a single tree with different separators? The design philosophy of tree-sitter grammars are that they are permissive: they accept syntax that is invalid, which works fine for their use case. So if someone wants to mix separators styles for whatever reason, that's fine, we can parse that.

Reply:
I also had this thought, but the PlusCal translator enforces the choice of a single syntax(either p- or c-), giving an error on a mixed-syntax spec.
One more thing to note here is that treesitter-based refactorings (like [these](https://github.com/ThePrimeagen/refactoring.nvim)) are not syntax-agnostic. I'm not sure, but it might be useful to know whether you are having braceless syntax or not

Reply:
Adding extra rules increases the (already large) size of the generated parser, which is a concern.

Reply:
We might glue the two syntaxes, but it would result in correct parser outputs on invalid specs. Is this what we are ready to go for?

Reply:
I'm totally fine with accepting invalid syntax (we already do this with some jlist stuff). But if the size difference isn't large then sure we can keep them around. Although a secondary concern is that it makes queries less general; you have to write separate queries for C syntax and P syntax.

Reply:
I've got these numbers for the `tlaplus.so` binary
|Revision|Size (bytes) |Diff with previous (bytes)|
|---|---|---|
|TLA+ only| 3910088 | -|
|TLA+, PlusCal (P) |4761232 | 851144| 
|TLA+, PlusCal(P, C)| 4905584| 144352 |

Reply:
What about the generated parser.c file?

Reply:
Updated the table with parser.c sizes

Reply:
Been thinking about this and am leaning toward combining the trees. Pros:
- Don't need to duplicate pluscal queries; better for usability of this grammar
- Grammar file is less confusing for future developers, with fewer rules
- Avoid common problems with code duplication like fixes to one not being made to the other
- Slightly decrease size of generated C code and resulting binary (tree-sitter combines the rule trees internally)

Cons:
- We admit some invalid syntax, which isn't really a problem because we can't surface the error to the user anyway

What do you think?

Reply:
I agree with the pros you've listed. However, I find it quite handy when a missing highlighting hints that  a code is not syntactically correct.

We have one more option here: aliasing (e.g. referencing both `pcal_p_if` and `pcal_c_if` as `pcal_if`).
It would address the issue of duplicate queries without giving up on accurate parsing.

Also I've already tried to reuse common rules as much as possible (e.g. [`pcal_algorithm_start`](https://github.com/susliko/tree-sitter-tlaplus/blob/6b68bde939faf9b87aeca673a9888c2ada11f2a3/grammar.js#L1244)), leaving only syntax-specific stuff like parenthesis.

One more thing to note is that merging two syntaxes is unlikely to make the grammar easier for reading, since it won't reduce the amount of code while increasing the rules complexity.

How do you like the approach of aliasing?

Reply:
Aliasing sounds like a good compromise!